onnxruntime/core/framework/execution_provider.cc

```diff
@@ -52,103 +52,52 @@
-    const int key = MakeKey(mem_info.id, mem_info.mem_type);
-    allocators[key] = replacing_allocator;
-    return true;
-  }
-
-  return false;
-}
-
 // Update allocator in the provider if already present; ignore if not.
+// We match using the device id, OrtMemType and OrtDevice info.
+// We ignore the allocator name, and OrtAllocatorType (whether internally an arena is used or not).
+// TODO: We should remove OrtAllocatorType from OrtMemoryInfo as it's an implementation detail of the allocator.
 void IExecutionProvider::ReplaceAllocator(AllocatorPtr allocator) {
   const auto& info = allocator->Info();
 
-  if (FindAndReplaceAllocator(info, mem_info_set_, allocators_, allocator)) {
-    // We found an allocator corresponding to the provided
-    // allocator's OrtMemoryInfo and we replaced it with the
-    // provided allocator.
-    // We return back.
-    return;
-  }
-
-  else {
-    // If we can't find an allocator registered with the exact OrtMemoryInfo
-    // as that of the replacing allocator, we do a "loosened" check
-    // (i.e.) check if there is an allocator registered with OrtAllocatorType
-    // as OrtArenaAllocator because for external user provided allocator
-    // we only accept OrtAllocatorType as OrtDeviceAllocator.
-    // If we do find such a registered allocator, we can safely go ahead
-    // and replace that with the provided allocator. This may seem like
-    // we are replacing an arena allocator with a non-arena allocator
-    // but in reality any user provided allocator may still be an arena
-    // allocator. We don't allow users to use OrtAllocatorType as
-    // OrtArenaAllocator for their allocators because we reserve its usage
-    // for our internal BFCArena.
-    // TODO: Should we remove the OrtAllocatorType field from OrtMemoryInfo to
-    // avoid such problems and also remove the unintuitive phenomenon of binding
-    // the allocator type info to OrtMemoryInfo (which loosely is just device info) ?
-    const auto& original_info = allocator->Info();
-
-    // If the alloc_type was OrtArenaAllocator already, then it is a no-op
-    if (original_info.alloc_type == OrtAllocatorType::OrtArenaAllocator) {
-      return;
-    }
-
-    auto check_info = original_info;
-
-    // Mutate the alloc_type
-    check_info.alloc_type = OrtAllocatorType::OrtArenaAllocator;
-
-    if (FindAndReplaceAllocator(check_info, mem_info_set_,
-                                allocators_, allocator)) {
-      // We found an allocator corresponding to the mutated OrtMemoryInfo
-      // and we replaced it with the provided allocator.
-      // Before we return back, we need to do some house-keeping
-      // (i.e.) update the EP's OrtMemoryInfo set
-
-      // Delete the existing OrtMemoryInfo  corresponding to the allocator
-      // that was replaced
-      mem_info_set_.erase(check_info);
+  // TODO: This only works on allocators that are stored in this class. If a derived class overrides GetAllocator
```

Comment:
>If a derived class overrides GetAllocato [](http://example.com/codeflow?start=72&length=40)

Should we make a comment about it on the class level or even make the class final?